tests/cbmc/proofs/s2n_hash_free/s2n_hash_free_harness.c

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "api/s2n.h"
+#include "crypto/s2n_hash.h"
+
+#include <cbmc_proof/make_common_datastructures.h>
+#include <cbmc_proof/proof_allocators.h>
+
+void s2n_hash_free_harness()
+{
+    /* Non-deterministic inputs. */
+    struct s2n_hash_state *state = cbmc_allocate_s2n_hash_state();
+
+    /* Assumptions. */
+    __CPROVER_assume(IMPLIES(state != NULL, s2n_hash_state_is_valid(state)));
+
+    /* Operation under verification. */
+    if (s2n_hash_free(state) == S2N_SUCCESS && state != NULL)
+    {
+        /* Post-conditions. */
+        assert(IMPLIES(!s2n_is_in_fips_mode(), state->is_ready_for_input == 0));
+        assert(IMPLIES(s2n_is_in_fips_mode(), state->digest.high_level.evp.ctx == NULL));
+        assert(IMPLIES(s2n_is_in_fips_mode(), state->digest.high_level.evp_md5_secondary.ctx == NULL));
+        assert(IMPLIES(s2n_is_in_fips_mode(), state->is_ready_for_input == 0));
```

Comment:
Doesn't this rely on the implementation of another method (s2n_hash_set_impl) to set the hash implementations the way you expect? Could you explicitly check state->hash_impl instead?

Reply:
@lrstewart In this case, if `s2n_is_in_fips_mode()` is true, the `state->hash_impl` is `s2n_evp_hash`.

```c
static int s2n_hash_set_impl(struct s2n_hash_state *state)
{
    state->hash_impl = s2n_is_in_fips_mode() ? &s2n_evp_hash : &s2n_low_level_hash;

    return S2N_SUCCESS;
}
```

I wouldn't be able to write the following either, because `s2n_low_level_hash` is statically defined in `s2n_hash.c`
```c
        assert(IMPLIES(state->hash_impl == s2n_low_level_hash, state->is_ready_for_input == 0));
```

Reply:
Fair. I really wish you didn't have to do it this way, but I can't figure out another way to do it :(